src/client/common/alias_dict.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/cli/alias_dict.h>
+#include <multipass/constants.h>
+#include <multipass/standard_paths.h>
+
+#include <QDir>
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonParseError>
+#include <QTemporaryFile>
+
+#include "../../daemon/json_writer.h"
+
+mp::AliasDict::AliasDict()
+{
+    const auto file_name = QStringLiteral("%1_aliases.json").arg(mp::client_name);
+    const auto user_config_path = QDir{MP_STDPATHS.writableLocation(mp::StandardPaths::GenericConfigLocation)};
+    const auto cli_client_dir_path = QDir{user_config_path.absoluteFilePath(mp::client_name)};
+
+    aliases_file = cli_client_dir_path.absoluteFilePath(file_name).toStdString();
+
+    load_dict();
+}
+
+mp::AliasDict::~AliasDict()
+{
+    if (modified)
+    {
+        save_dict();
+    }
+}
+
+void mp::AliasDict::add_alias(const std::string& alias, const mp::AliasDefinition& command)
+{
+    if (aliases.try_emplace(alias, command).second)
+    {
+        modified = true;
+    }
+}
+
+bool mp::AliasDict::remove_alias(const std::string& alias)
+{
+    if (aliases.erase(alias) > 0)
+    {
+        modified = true;
+        return true;
+    }
+
+    return false;
+}
+
+mp::AliasDict::size_type mp::AliasDict::remove_aliases_for_instance(const std::string& instance)
```

Comment:
I'm not sure why this returns a size as it is not used by any of the callers.  You could then also tighten up the code below by getting rid of `erased_elements`, setting `modified = true` in the loop (assigning is a boolean is less expensive than a `++` operation, not that we really care about that :wink: ), and get rid of the `if (erased_elements > 0)` logic.

Reply:
I thought it'll be useful to know how many aliases were removed but you're right: it's useless. Changed as you suggested: made the function `void` because the only thing that matters is the `modified` flag.